deepola/wake/src/polars_operations/merger.rs

```diff
@@ -0,0 +1,569 @@
+impl<T, P: MergerOp<T> + Clone> Default for MergerNode<T, P> {
+    fn default() -> Self {
+        Self {
+            merger: P::new(),
+            phantom: PhantomData::default(),
+        }
+    }
+}
+
+impl<T: 'static + Send, P> MergerNode<T, P>
+where
+    P: 'static + MergerOp<T> + StreamProcessor<T> + Clone,
+{
+    pub fn new() -> Self {
+        MergerNode::default()
+    }
+
+    pub fn merger(&mut self, op: P) -> &mut Self {
+        self.merger = op;
+        self
+    }
+
+    pub fn build(&self) -> ExecutionNode<T> {
+        let data_processor = Box::new(self.merger.clone());
+        ExecutionNode::<T>::new(data_processor, 2)
+    }
+}
+
+/// Represents a generic merger operation. For [MergerNode], a concrete struct implementing this
+/// trait (e.g., SortedDfMerger) must be used.
+pub trait MergerOp<T>: Send {
+    fn new() -> Self;
+
+    fn supply_left(&self, df_left: &T);
+
+    fn supply_right(&self, df_right: &T);
+
+    fn needs_left(&self) -> bool;
+
+    fn needs_right(&self) -> bool;
+
+    fn merge(&self) -> T;
+}
+
+/// Inner-joins two dataframes, both of which sorted on their join keys. Assumes that the next data 
+/// frame (each for left and right) does not contain the same key values appearing in the current 
+/// data frame.
+#[derive(Getters, Setters, Clone)]
+pub struct SortedDfMerger {
+    // left join key
+    #[set = "pub"]
+    #[get = "pub"]
+    left_on: Vec<String>,
+
+    // right join key
+    #[set = "pub"]
+    #[get = "pub"]
+    right_on: Vec<String>,
+
+    // will be used v2
+    left_df: RefCell<Option<DataFrame>>,
+
+    // will be used v2
+    right_df: RefCell<Option<DataFrame>>,
+
+    needs_left: RefCell<bool>,
+
+    needs_right: RefCell<bool>,
+}
+
+impl SortedDfMerger {
+    fn new() -> Self {
+        SortedDfMerger {
+            left_on: Vec::new(),
+            right_on: Vec::new(),
+            left_df: RefCell::new(None),
+            right_df: RefCell::new(None),
+            needs_left: RefCell::new(true),
+            needs_right: RefCell::new(true),
+        }
+    }
+
+    /// Polars library has a bug that if dataframes to join are empty, it panics.
+    fn inner_join_possibly_empty(&self, df_left: &DataFrame, df_right: &DataFrame) -> DataFrame {
+        if df_left.height() == 0 || df_right.height() == 0 {
+            let mut new_columns = Vec::<Series>::new();
+            // push all the columns from the left
+            for left_col in df_left.get_columns() {
+                new_columns.push(left_col.clone());
+            }
+            // push all the columns from the right, except for the ones composing join cols
+            for right_col in df_right.get_columns() {
+                if self.right_on.contains(&String::from(right_col.name())) {
+                    continue;
+                }
+                new_columns.push(right_col.clone());
+            }
+            DataFrame::new(new_columns).unwrap()
```

Comment:
What is the idea behind this? If either one's height is 0, shouldn't the join output be empty? Also, doesn't this mean that `new_columns` will have Series of varying length (the non-zero length as well as the 0 length ones) - can a DataFrame be constructed using that?

Reply:
If either one's height is 0, Polars complains (instead of producing an empty dataframe).

You are right. I should have added only zero-size Series. I will make a change accordingly.

Reply:
@nikhil96sher I just made a change and two related test cases.